src/NuGet.Core/NuGet.Protocol.Core.v3/HttpSource/HttpHandlerResourceV3Provider.cs

```diff
@@ -105,17 +116,44 @@ public CredentialPromptWebRequestHandler(IWebProxy proxy, ICredentials credentia
                         if (ProxyAuthenticationRequired(ex) &&
                             HttpHandlerResourceV3.PromptForProxyCredentials != null)
                         {
-                            // prompt use for proxy credentials.
-                            var credentials = await HttpHandlerResourceV3
-                                .PromptForProxyCredentials(request.RequestUri, Proxy, cancellationToken);
+                            ICredentials currentCredentials = Proxy.Credentials;
 
-                            if (credentials == null)
+                            try
                             {
-                                throw;
+                                await _credentialPromptLock.WaitAsync();
+
+                                // Check if the credentials have already changed
+                                if (beforeAuthId != _lastAuthId)
+                                {
+                                    continue;
+                                }
+
+                                // Limit the number of retries
+                                _authRetries++;
+                                if (_authRetries >= MaxAuthRetries)
```

Comment:
The retry on the auth prompt involves a user entering (hopefully) different info each time. The retry for failed requests involves sending the same exact request each time.

The auth prompt retry number here is a max, users can exit it on the first pass by not entering any credentials if I remember right.

For network errors the retry is kept low so that a failing server does not cause an operation to go on for extended period of time when it has little chance of succeeding.

The law of diminishing returns applies to both all scenarios, however I would say it diminishes faster in some scenarios than others. For example 5xx in practice is often a more temporary issue than other failures which is why NuGet handles it separately. Going through a list of multiple passwords, possible casings, or variations which may have been used on the feed may not be diminishing at all until the user gives up.